test/passing/break_cases_sparse.ml

```diff
@@ -0,0 +1,68 @@
+let f x = function
+  | C
+  | P (this, test, [is; wide; enough; _to; break], [the; line])
+  | A
+  | K ->
+      1
+  | D ->
+      let a = "this" in
+      let b = "breaks" in
+      ()
+
+
+let f =
+  let g = function
+    | H
+      when x y <> k ->
+        2
+    | T
+    | P
+    | U ->
+        3
+  in
+  fun x g t h y u ->
+    match x with
+    | E ->
+        4
+    | Z
+    | P
+    | M ->
+      match y with
+      | O ->
+          5
+      | P
+        when h x ->
+          function
+          | A ->
+              6
+
+
+;;
+match x with
+| true -> (
+  match y with
+  | true ->
+      "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+  | false ->
+      "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" )
+| false ->
+    "cccccccccccccccccccccccccccccc"
```

Comment:
Looks good! Are you happy with this now and ready for this to be merged, or is there a plan to try to avoid breaking in cases like the line above while preserving the breaks like in cases below?

Reply:
If there was an easy way to avoid breaking in the case above I think it would look even better, but I did not manage to do that without breaking the other cases. If someone thinks it should be done in this PR (@samoht ?) I am open to suggestions.

Reply:
No opinion, as I will probably not use that sparse option :-)

Reply:
Actually, I am a bit lost. How do I get the tool to format:

```ocaml
let f x = function
  | C -> 1
  | D -> 2
```

without adding line breaks?

Reply:
Ok so I am very confused now.  `ocamlformat` always formats this into:

```ocaml
let f x = function
  | x ->
      1
  | y ->
      2
```

even without using `--sparse` option.

But maybe this should be done with an other option in a different PR...

Reply:
Yes I went back with the formatting of arrows because I did not manage to:
- break after `->` when the patterns or the when clause or the rhs expression breaks
- not break otherwise
When I tried to remove the break hint, I had results like:
```
let is_sequence exp =
  match exp.pexp_desc with
  | Pexp_sequence _
   |Pexp_extension
      ( _
      , PStr [{pstr_desc= Pstr_eval ({pexp_desc= Pexp_sequence _}, []); _}]
      ) -> true
```
`) -> true` fits on the line, but since the pattern breaks, we also would like to breka after `->`

So maybe it is better to stick with the default formatting of arrows for now.

Reply:
@samoht, the cause of your confusion is why I asked if this is ready, since as far as I understand it doesn't completely match the initial goal.

There is some interaction between the breaking after `->` of individual cases and the breaking between cases. The current patch force-breaks the box enclosing the cases, but that has the additional effect of also force-breaking after `->`. I haven't thought through it in detail, but I suspect that it is necessary to add an additional level (probably two, actually) of boxing around individual case patterns + `->` as well as around whole cases. That way the box enclosing the cases can be force-broken while letting the individual cases break naturally, with a box around the pattern and `->` so that if the pattern breaks, there is also a break after the `->`.

Reply:
I tried what you suggested but did not manage to make it work (tried several combinations of box types). I don't know if it is worth investigating.
By the way I squashed the commits in case you want to merge.